src/main/python/pybuilder/plugins/python/coverage_plugin.py

```diff
@@ -130,38 +142,87 @@ def do_coverage(project, logger, reactor, execution_prefix, execution_name, targ
+
+    for module in modules:
+        module_name = module.__name__
 
-        module_report_data = build_module_report(coverage, module)
-        should_ignore_module = module_name in exceptions
+        module_report_data = _build_module_report(coverage, module)
 
-        if not should_ignore_module:
-            sum_lines += module_report_data[0]
-            sum_lines_not_covered += module_report_data[2]
+        sum_lines += module_report_data.n_lines_total
+        sum_lines_not_covered += module_report_data.n_lines_missing
+        sum_branches += module_report_data.n_branches
+        sum_branches_missing += module_report_data.n_branches_missing
+        sum_branches_partial += module_report_data.n_branches_partial
 
         module_report = {
             "module": module_name,
-            "coverage": module_report_data[4],
-            "sum_lines": module_report_data[0],
-            "lines": module_report_data[1],
-            "sum_lines_not_covered": module_report_data[2],
-            "lines_not_covered": module_report_data[3],
+            "coverage": module_report_data.code_coverage,
+            "sum_lines": module_report_data.n_lines_total,
+            "lines": module_report_data.lines_total,
+            "sum_lines_not_covered": module_report_data.n_lines_missing,
+            "lines_not_covered": module_report_data.lines_missing,
+            "branches": module_report_data.n_branches,
+            "branches_partial": module_report_data.n_branches_partial,
+            "branches_missing": module_report_data.n_branches_missing
         }
 
         logger.debug("Module coverage report: %s", module_report)
         report["module_names"].append(module_report)
-        if module_report_data[4] < threshold:
-            msg = "Test coverage below %2d%% for %s: %2d%%" % (threshold, module_name, module_report_data[4])
-            if not should_ignore_module:
-                logger.warn(msg)
-                coverage_too_low = True
-            else:
-                logger.info(msg)
+        if module_report_data.code_coverage < threshold:
+            msg = "Test coverage below %2d%% for %s: %2d%%" % (threshold, module_name, module_report_data.code_coverage)
+            logger.warn(msg)
+            coverage_too_low = True
+        if module_report_data.branch_coverage < branch_threshold:
+            msg = "Branch coverage below %2d%% for %s: %2d%%" % (
+                branch_threshold, module_name, module_report_data.branch_coverage)
+            logger.warn(msg)
+            branch_coverage_too_low = True
+
+        if module_report_data.branch_partial_coverage < branch_partial_threshold:
+            msg = "Partial branch coverage below %2d%% for %s: %2d%%" % (
+                branch_partial_threshold, module_name, module_report_data.branch_partial_coverage)
+            logger.warn(msg)
+            branch_partial_coverage_too_low = True
 
     if sum_lines == 0:
         overall_coverage = 0
```

Comment:
@mriehl when there is no code we want to show no coverage? I would have assumed that 0 lines are always 100% covered. I followed that logic elsewhere as well, but I'm not sure it should be kept. Thoughts?